apps/namadillo/src/App/Transfer/DestinationAddressModal.tsx

```diff
@@ -0,0 +1,286 @@
+    (account) => account.type !== AccountType.ShieldedKeys
+  );
+  const shieldedAccount = accounts?.find(
+    (account) => account.type === AccountType.ShieldedKeys
+  );
+
+  // Dont display an address if it matches the source address
+  const isSourceAddressMatch = (address: string): boolean =>
+    address === sourceAddress;
+
+  // Build your addresses options
+  const addressOptions: AddressOption[] = [];
+  if (accounts) {
+    const transparentAccount = accounts.find(
+      (account) => account.type !== AccountType.ShieldedKeys
+    );
+
+    if (
+      transparentAccount &&
+      !isSourceAddressMatch(transparentAccount.address)
+    ) {
+      addressOptions.push({
+        id: "transparent",
+        label: "Transparent Address",
+        address: transparentAccount.address,
+        icon: namadaTransparentIcon,
+        type: "transparent",
+      });
+    }
+    if (shieldedAccount && !isSourceAddressMatch(shieldedAccount.address)) {
+      addressOptions.push({
+        id: "shielded",
+        label: "Shielded Address",
+        address: shieldedAccount.address,
+        icon: namadaShieldedIcon,
+        type: "shielded",
+      });
+    }
+  }
+  if (keplrAddress)
+    addressOptions.push({
+      id: "keplr",
+      label: "Keplr Address",
+      address: keplrAddress ?? "",
+      icon:
+        !keplrAddress ?
+          wallets.keplr.iconUrl
+        : getChainImageUrl(getChainFromAddress(keplrAddress ?? "")),
+      type: "keplr",
+    });
+
+  // Build recent addresses options
+  const recentAddressOptions: AddressOption[] = recentAddresses.map(
+    (recent) => ({
+      id: `recent-${recent.address}`,
+      label: recent.label || getAddressLabel(recent.address, recent.type),
+      address: recent.address,
+      icon:
+        recent.type === "shielded" ? namadaShieldedIcon
+        : recent.type === "transparent" ? namadaTransparentIcon
+        : getChainImageUrl(getChainFromAddress(recent.address ?? "")), // fallback for IBC
+      type: recent.type,
+    })
+  );
+
+  const handleAddressClick = (address: string): void => {
+    onSelectAddress(address);
+    onClose();
+  };
+
+  const handleCustomAddressChange = (
+    e: React.ChangeEvent<HTMLInputElement>
+  ): void => {
+    const value = e.target.value;
+    setCustomAddress(value);
+
+    // Validate the address as user types
+    if (value.trim()) {
+      const result = validateAddress(value);
+      setValidationResult(result);
+    } else {
+      setValidationResult(null);
+    }
+  };
+
+  const handleCustomAddressSubmit = (): void => {
+    const trimmedAddress = customAddress.trim();
+    if (!trimmedAddress) return;
+
+    const result = validateAddress(trimmedAddress);
+    setValidationResult(result);
+
+    if (result.isValid && result.addressType) {
+      // Add to recent addresses
+      const label = getAddressLabel(trimmedAddress, result.addressType);
+      const updatedRecents = addToRecentAddresses(
```

Comment:
Side note, but it would be cool to have an ability to remove addresses from recent ones :)